ports/rp2/main.c

```diff
@@ -166,6 +172,9 @@ int main(int argc, char **argv) {
         mod_network_lwip_init();
         #endif
 
+        // Hook for setting up anything that should follow the MicroPy runtime but be prior to _boot.py
+        MICROPY_BOARD_LATE_INIT();
```

Comment:
From a maintainability perspective, I have some concerns about having different board-level hooks on different ports that implement similar or the same functionality.

The way this is done on the stm32 & renesas-ra ports is there is a hook `MICROPY_BOARD_RUN_BOOT_PY(&state)` which calls `boardctrl_run_boot_py()` by default, but can be hooked to do something else instead.

I'm not sure that's necessarily a better design, but I am sure that doing it differently per-port will eventually become a pain to maintain.

How would you feel about refactoring the boot.py logic immediately below this into a separate function `rp2_run_boot_py` or similar, and have his hook call there by default or optionally be overridden?

Reply:
As mentioned in my main reply, I'm not actively using this hook, so it could just be removed for the purpose of getting this PR through.

That being said, I do patch the Pimoroni MicroPython build for Yukon with a similar `state` for determining if boot and main py should run, and what you propose would let me do this in an official way without the need to keep a fork active, if I am understanding correctly.

```c
    if (run_main) {
            // Execute user scripts.
            int ret = pyexec_file_if_exists("boot.py");
            if (ret & PYEXEC_FORCED_EXIT) {
                run_main = !(ret & PYEXEC_SKIP_USER_CODE);
                goto soft_reset_exit;
            }
            if (pyexec_mode_kind == PYEXEC_MODE_FRIENDLY_REPL) {
                ret = pyexec_file_if_exists("main.py");
                if (ret & PYEXEC_FORCED_EXIT) {
                    run_main = !(ret & PYEXEC_SKIP_USER_CODE);
                    goto soft_reset_exit;
                }
            }
        }
        run_main = true;
```

My only hesitation with me doing what you propose (versus someone more familiar with the codebase) is the bigger impact it may have on other RP2 products, since I am likely to break things.

Reply:
> what you propose would let me do this in an official way without the need to keep a fork active, if I am understanding correctly.

I think so, yes.

> is the bigger impact it may have on other RP2 products, since I am likely to break things.

I think the chance of breakage is pretty low, provided the default implementation of the hook calls a function that perform the "Execute user scripts" stage as it's implemented now. In that case the only change is really the additional function call and maybe a return check. I'm very happy to review such a change.

For a reference of the "pattern" for such an implementation, the stm32 port is a good reference:
https://github.com/micropython/micropython/blob/master/ports/stm32/main.c#L578
https://github.com/micropython/micropython/blob/master/ports/stm32/boardctrl.h#L64
https://github.com/micropython/micropython/blob/master/ports/stm32/boardctrl.c#L192

This raises the question of whether rp2 port should add its own boardctrl.[hc] as well. Thinking about it now, I feel this might be a good refactoring to keep these new features well-defined, even if it's a pretty minimal implementation. However I'm open to other approaches, too. :)

All that said, if you don't actually need this right now then pulling it out of this PR and potentially doing it in a future PR is also an option.

Reply:
Okay. I've updated this PR to only have 4 hooks, and renamed them to match the STM32's naming scheme.

Although it would be nice to have the boot.py and main.py stages follow the pattern used by STM32, at this time I do not need that level of control for Yukon. Turns out that code I showed that modified the main run process never actually did anything different as the `PYEXEC_SKIP_USER_CODE` value I added was never actually emitted by anything.